nexus/db-model/src/multicast_group.rs

```diff
@@ -0,0 +1,483 @@
+pub struct IncompleteExternalMulticastGroupParams {
+    pub id: Uuid,
+    pub name: Name,
+    pub description: String,
+    pub ip_pool_id: Uuid,
+    pub rack_id: Uuid,
+    pub explicit_address: Option<IpAddr>,
+    pub source_ips: Vec<IpNetwork>,
+    pub mvlan: Option<i16>,
+    pub vni: Vni,
+    pub tag: Option<String>,
+}
+
+impl IncompleteExternalMulticastGroup {
+    /// Create an incomplete multicast group from parameters.
+    pub fn new(params: IncompleteExternalMulticastGroupParams) -> Self {
+        Self {
+            id: params.id,
+            name: params.name,
+            description: params.description,
+            time_created: Utc::now(),
+            ip_pool_id: params.ip_pool_id,
+            source_ips: params.source_ips,
+            explicit_address: params.explicit_address.map(|ip| ip.into()),
+            mvlan: params.mvlan,
+            vni: params.vni,
+            tag: params.tag,
+            rack_id: params.rack_id,
+        }
+    }
+}
+
+impl MulticastGroupMember {
+    /// Generate a new multicast group member.
+    ///
+    /// Note: version_added will be set by the database sequence when inserted.
+    pub fn new(
+        id: Uuid,
+        external_group_id: Uuid,
+        parent_id: Uuid,
+        sled_id: Option<DbTypedUuid<SledKind>>,
+    ) -> Self {
+        Self {
+            id,
+            time_created: Utc::now(),
+            time_modified: Utc::now(),
+            time_deleted: None,
+            external_group_id,
+            parent_id,
+            sled_id,
+            state: MulticastGroupMemberState::Joining,
+            // Placeholder - will be overwritten by database sequence on insert
+            version_added: Generation::new(),
+            version_removed: None,
+        }
+    }
+}
+
+/// Database representation of an underlay multicast group.
+///
+/// Underlay groups are system-generated admin-scoped IPv6 multicast addresses
+/// used as a NAT target for internal multicast traffic.
+///
+/// These are distinct from [ExternalMulticastGroup] which are external-facing
+/// addresses allocated from IP pools, specified by users or applications.
+#[derive(
+    Queryable,
+    Insertable,
+    Selectable,
+    Clone,
+    Debug,
+    PartialEq,
+    Eq,
+    Serialize,
+    Deserialize,
+)]
+#[diesel(table_name = underlay_multicast_group)]
+pub struct UnderlayMulticastGroup {
+    /// Unique identifier for this underlay multicast group.
+    pub id: Uuid,
+    /// Timestamp for creation of this underlay multicast group.
+    pub time_created: DateTime<Utc>,
+    /// Timestamp for last modification of this underlay multicast group.
+    pub time_modified: DateTime<Utc>,
+    /// Timestamp for deletion of this underlay multicast group, if applicable.
+    pub time_deleted: Option<DateTime<Utc>>,
+    /// Admin-scoped IPv6 multicast address (NAT target).
+    pub multicast_ip: IpNetwork,
+    /// VNI for this multicast group.
+    pub vni: Vni,
```

Comment:
It's a bit odd to have a VNI associated with an underlay address. We also have this information in the overlay `MulticastGroup` data structure above (which I would expect). Why do we need it in both places? I typically think of the underlay addresses as a target that is VNI agnostic, or put differently, the VNI is a numeric identifier for an overlay.